tensorflow_lite_support/python/task/vision/pybinds/_pywrap_image_segmenter.cc

```diff
@@ -0,0 +1,79 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "pybind11/pybind11.h"
+#include "pybind11_protobuf/native_proto_caster.h"  // from @pybind11_protobuf
+#include "tensorflow_lite_support/cc/task/processor/proto/segmentations.pb.h"
+#include "tensorflow_lite_support/cc/task/processor/proto/segmentation_options.pb.h"
+#include "tensorflow_lite_support/cc/task/vision/image_segmenter.h"
+#include "tensorflow_lite_support/examples/task/vision/desktop/utils/image_utils.h"
+#include "tensorflow_lite_support/python/task/core/pybinds/task_utils.h"
+
+namespace tflite {
+namespace task {
+namespace vision {
+
+namespace {
+namespace py = ::pybind11;
+using PythonBaseOptions = ::tflite::python::task::core::BaseOptions;
+using CppBaseOptions = ::tflite::task::core::BaseOptions;
+}  // namespace
+
+PYBIND11_MODULE(_pywrap_image_segmenter, m) {
+  // python wrapper for C++ ImageSegmenter class which shouldn't be directly
+  // used by the users.
+  pybind11_protobuf::ImportNativeProtoCasters();
+
+  py::class_<ImageSegmenter>(m, "ImageSegmenter")
+      .def_static(
+          "create_from_options",
+          [](const PythonBaseOptions& base_options,
+             const processor::SegmentationOptions& segmentation_options) {
+            ImageSegmenterOptions options;
+            auto cpp_base_options =
+                core::convert_to_cpp_base_options(base_options);
+            options.set_allocated_base_options(cpp_base_options.release());
+
+            if (segmentation_options.has_display_names_locale()) {
+              options.set_display_names_locale(
+            segmentation_options.display_names_locale());
+            }
+            if (segmentation_options.has_output_type()) {
+              options.set_output_type(
+                  static_cast<ImageSegmenterOptions::OutputType>(
+                          segmentation_options.output_type()));
+            }
+
+            auto segmenter = ImageSegmenter::CreateFromOptions(options);
+            return core::get_value(segmenter);
+          })
+      .def("segment",
+           [](ImageSegmenter& self, const ImageData& image_data)
+               -> processor::SegmentationResult {
+             auto frame_buffer = CreateFrameBufferFromImageData(image_data);
+             auto vision_segmentation_result = self.Segment(
+                     *core::get_value(frame_buffer));
+             // Convert from vision::SegmentationResult to
+             // processor::SegmentationResult as required by the Python layer.
+             processor::SegmentationResult segmentation_result;
```

Comment:
@kinaryml As the segmentation result contains large object, we shouldn't serialize it because it's expected to be slow. 

Instead, could you run a benchmark to compare these approach?
* Serialize then Parse (the current approach)
* Memory copy the mask(s) but serialize and parse the ColorLabels
* Pass reference of the mask(s) but serialize and parse the ColorLabels

Please try with a mask of size 512 * 512 and we'll see how difference the performance are and decide what should we use.

Reply:
@khanhlvg Could you tell me where I can find the 512x512 mask? I'm not sure if I can benchmark it FYI. 

On a side note, wouldn't migrating the return types of the Image C++ Task APIs to use the processor `segmentations` proto do away with this step altogether? I believe some of the other APIs like Audio and Text use return types from the `processor` namespace instead of their own task's namespace. This would avoid parsing and converting the results in `ImageClassifier` and similar image tasks and return the underlying data directly like we do in these instances.

Audio
[EmbeddingResult](https://github.com/khanhlvg/tflite-support/blob/77725772ea060da2ac8964ee1216350e9c6ef6e1/tensorflow_lite_support/python/task/audio/pybinds/_pywrap_audio_embedder.cc#L62)
[ClassificationResult](https://github.com/khanhlvg/tflite-support/blob/77725772ea060da2ac8964ee1216350e9c6ef6e1/tensorflow_lite_support/python/task/audio/pybinds/_pywrap_audio_classifier.cc#L71)

Text
[EmbeddingResult](https://github.com/khanhlvg/tflite-support/blob/77725772ea060da2ac8964ee1216350e9c6ef6e1/tensorflow_lite_support/python/task/text/pybinds/_pywrap_text_embedder.cc#L53)

Reply:
@kinaryml Migrating the C++ layer requires a lot of effort because it's used by many dependencies (e.g. Java, iOS, C API and other internal dependencies that aren't on GitHub). Therefore we don't want to migrate now.

To run the benchmark, you should write a new C code that does these things:
* Create a vision::SegmentationResult with 21 random confidence mask of size 512 * 512.
* Convert the vision::SegmentationResult object to a processor::SegmentationResult following the 3 approaches I mentioned earlier.
* Measure the time needed for the conversion. You should run the conversion for 100 times and calculate the average.

I'd expect a processing time report like [this](https://github.com/tensorflow/tflite-support/pull/752#discussion_r819346296). From there, let's decided which approach we should take.

Reply:
@khanhlvg I'm not quite sure how to setup and run the benchmark reliably in C++ FYI as I'm not entirely familiar using protobuf with C++ in mind. Would it possible for anyone on your end to investigate and decide which approach works better and we could apply it here.

Reply:
@kinaryml I'll look into how to implement a more efficient copy later. Now let's use the Serialization as what we're doing now.